### PR TITLE
Revoke Twitter connection when the token is revoked

### LINKED ISSUE #123: Remove connection to a social service if the token expired
Every now and then I see errors like `Twitter::Error::Unauthorized: Invalid or expired token` whenever someone answers a question.

It would be nice to respond to unauthorised errors with the disconnection of a service.  Ideally the user should be informed about that, preferably not via a question (related: #78).

### PR DESCRIPTION
**Pull request by raccube**
Closes #123

### CODE REVIEW

**Review by pixeldesu — APPROVED** [review]

File: app/workers/share_worker.rb [hunk 1]
```diff
@@ -9,8 +9,7 @@ class ShareWorker
   # @param answer_id [Integer] the user id
   # @param service [String] the service to post to
   def perform(user_id, answer_id, service)
-    service_type = "Services::#{service.camelize}"
-    user_service = User.find(user_id).services.find_by(type: service_type)
+    user_service = find_service(user_id, service)
```
Comment (nilsding) [inline app/workers/share_worker.rb]: could use an instance variable here

```suggestion
    @user_service = User.find(user_id).services.find_by(type: "Services::#{service.camelize}")
```

File: app/workers/share_worker.rb [hunk 2]
```diff
@@ -23,11 +22,25 @@ def perform(user_id, answer_id, service)
...
+    revoke_and_notify(user_id, service)
   rescue => e
     logger.info "failed to post answer #{answer_id} to #{service} for user #{user_id}: #{e.message}"
     Sentry.capture_exception(e)
     raise
   end
+
+  def revoke_and_notify(user_id, service)
+    user_service = find_service(user_id, service)
+    user_service.destroy
```
Comment (nilsding) [inline app/workers/share_worker.rb]: ...which can then be used here to destroy the service

```suggestion
    @user_service.destroy
```

-> 1 query less!

File: app/workers/share_worker.rb [hunk 3]
```diff
@@ -23,11 +22,25 @@ def perform(user_id, answer_id, service)
...
+
+    Object.const_get("Notification::ServiceTokenExpired").create(
+      target_id:    user_id,
+      target_type:  "User::Expired#{service.camelize}ServiceConnection",
+      recipient_id: user_id,
+      new:          true
+    )
+  end
+
+  def find_service(user_id, service) = User.find(user_id).services.find_by(type: "Services::#{service.camelize}")
```
Comment (nilsding) [inline app/workers/share_worker.rb]: (and then get rid of that method)

```suggestion
```

File: app/workers/share_worker.rb [hunk 4]
```diff
@@ -9,8 +9,7 @@ class ShareWorker
   # @param answer_id [Integer] the user id
   # @param service [String] the service to post to
   def perform(user_id, answer_id, service)
-    service_type = "Services::#{service.camelize}"
-    user_service = User.find(user_id).services.find_by(type: service_type)
+    user_service = find_service(user_id, service)
 
     user_service.post(Answer.find(answer_id))
```
Comment (nilsding) [inline app/workers/share_worker.rb]:
```suggestion
    @user_service.post(Answer.find(answer_id))
```

**Review by nilsding — CHANGES_REQUESTED** [review]

**Review by nilsding — APPROVED** [review]

**Pull request merged**